Enable isolation of a single fenced block example

cef62 opened the pull request:
Hi @sapegin 
first of all thanks for this project! I really love it and we use it in several projects.
In many of our library projects is the only available compiled output  and we often need to isolate a single example from a readme file.

This PR add the ability to isolate a single code example to simplify the development process.
Would you like to add this feature to the official repo? 

I'd love to have your feedback!
Cheers

sapegin commented:
Hi @cef62! That looks very interesting, thank you very much for the PR!

I’ll review the code a bit later.

Using indexes looks fragile (what if you add or move an example?) but I have no better idea how to implement it.

cef62 commented:
Thanks @sapegin, let me knows you want me to change something!

> Using indexes looks fragile (what if you add or move an example?) but I have no better idea how to implement it.

I agree with you but I didn't want to touch too much your current code structure and all texts and fenced blocks are stored in a single array with just numeric indexes. Anyway I don't think should be a real problem because the index will change only if you add other texts or fenced blocks to the readme file and that shouldn't happen while you are debugging a single example. 
Probably the only problematic scenario is if you bookmark a single example and then the readme file changes. 

I think I can get a more robust solution but that would involve the user somehow to explicitly always specify an ID for the fenced block; personally, as user, I'd prefer avoiding it but it's a viable option.

sapegin commented:
I think indexes are fine unless someone find better solution that doesn’t affect developer experience.

sapegin commented on src/index.js:
```diff
@@ -25,15 +26,21 @@ function renderStyleguide() {
 		sections = [];
 		sidebar = false;
+
+		if (components.length === 1 && _.isFinite(targetComponentIndex)) {
```
Please use cherry-picking for Lodash for consistency: `import isFinite from 'lodash/isFinite'`.

sapegin replied:
Looks like we need some comments here to explain all the things we’re doing: isolated component, isolated example, etc.

sapegin commented on src/rsg-components/Playground/Playground.css:
```diff
@@ -14,7 +14,7 @@
 	composes: border link from "../../styles/colors.css";
 	position: absolute;
-	right: -1px;
+	right: 138px;
```
This is not cool ;-)

sapegin commented on src/index.js:
```diff
@@ -43,6 +50,7 @@ function renderStyleguide() {
 			components={components}
 			sections={sections}
 			sidebar={sidebar}
+			singleExample={singleExample}
```
I thinks it would be better to use context here, like for `config`. Probably make a new object in config for `sidebar` and `singleExample`.

cef62 commented:
> We need to do something with the UI. Right now it’s inconsistent and a bit too loud for such a feature (But I we can merge it and think more later.)

Yep the UI is not that good... my intention was to understand if you'd like the feature before thinking on the look & feel. 

> Otherwise it’s great ;-)
Thanks :)

I've addressed your comments, let me know if you want further tweaks.

cef62 commented:
I was thinking to move the link to isolate a single example to the upper-right of the rendered fenced block. Something similar the link you already have in place in the upper-right corner of every rendered component.

What do you think? Would you like I add it to the PR?

sapegin commented:
I was thinking about that too. And show it on hover. At least it would be more consistent with component isolation UI. So please add it to the PR and I’ll merge it.

cef62 commented:
Great! I'll add it to the PR.

cef62 commented:
I've made the discussed changes, let me know if it's ok for you!

sapegin commented:
Cool, that’s much better!

cef62 commented:
Thanks! I'm happy you liked the addition 👍
